File: examples/java/order/src/main/java/dev/angzarr/examples/order/OrderState.java
package dev.angzarr.examples.order;

import java.util.List;
import java.util.ArrayList;

public record OrderState(
    String customerId,
    List<LineItem> items,
    int subtotalCents,
    int discountCents,
    int loyaltyPointsUsed,
    String paymentMethod,
    String paymentReference,
    String status
) {
    public static OrderState empty() {
        return new OrderState("", new ArrayList<>(), 0, 0, 0, "", "", "");
    }

    public boolean exists() {
        return !customerId.isEmpty();
    }

    public boolean isPending() {
        return "pending".equals(status);
    }

    public int totalCents() {
        return subtotalCents - discountCents;
    }

    public record LineItem(String productId, String name, int quantity, int unitPriceCents) {}
}

File: examples/kotlin/order/src/main/kotlin/dev/angzarr/examples/order/OrderState.kt
package dev.angzarr.examples.order

import examples.Domains.LineItem

data class OrderState(
    val customerId: String = "",
    val items: List<LineItem> = emptyList(),
    val subtotalCents: Int = 0,
    val discountCents: Int = 0,
    val loyaltyPointsUsed: Int = 0,
    val finalTotalCents: Int = 0,
    val paymentMethod: String = "",
    val status: String = ""
) {
    fun exists(): Boolean = customerId.isNotEmpty()
    fun isPendingPayment(): Boolean = status == "pending_payment"
    fun isPaid(): Boolean = status == "paid"
    fun isCompleted(): Boolean = status == "completed"
    fun isCancelled(): Boolean = status == "cancelled"

    companion object {
        fun empty(): OrderState = OrderState()
    }
}
